lms/djangoapps/courseware/courses.py

```diff
@@ -449,10 +449,44 @@ def get_course_syllabus_section(course, section_key):
     raise KeyError("Invalid about key " + str(section_key))
 
 
+class AccessLimitedCourseList(object):
+    def __init__(self, course_queryset, user, permission_name):
+        self.course_queryset = course_queryset
+        self._access_limited_courses = (c for c in self.course_queryset if has_access(user, permission_name, c))
+        self._count = None
+        self._course_cache = []
+
+    def __len__(self):
+        if self._count is None:
+            self._count = self.course_queryset.count()
+        return self._count
+
+    def __iter__(self):
+        for course in self._course_cache:
+            yield course
+        for course in self._access_limited_courses:
+            self._course_cache.append(course)
+            yield course
+
+    def __getitem__(self, index):
+        if isinstance(index, int):
+            while len(self._course_cache) < index:
+                self._course_cache.append(next(self._access_limited_courses))
+            return self._course_cache[index]
+        elif isinstance(index, slice):
+            if index.stop is not None:
```

Comment:
You might want a check for the slice features you won't be implementing: `[10:1:-1]`, for example.  I guess `[0:10:2]` would work...